mllib/src/main/scala/org/apache/spark/ml/feature/FuncTransformer.scala

```diff
@@ -0,0 +1,141 @@
+ * Note that FuncTransformer supports serialization via Scala ObjectOutputStream and may not
+ * guarantee save/load compatibility between different Scala version.
+ * @param func a custom UserDefinedFunction to map from inputCol to outputCol e.g.
+ *             udf { (i: Double) => i + 1 }. Only udf with one input is supported for now.
+ */
+@Experimental
+@Since("2.3.0")
+class FuncTransformer @Since("2.3.0") (
+    @Since("2.3.0") override val uid: String,
+    @Since("2.3.0") val func: UserDefinedFunction
+  ) extends Transformer with HasInputCol with HasOutputCol with DefaultParamsWritable {
+
+  @Since("2.3.0")
+  def this(func: UserDefinedFunction) = this(Identifiable.randomUID("FuncTransformer"), func)
+
+  setDefault(inputCol -> "input", outputCol -> "output")
+
+  /** @group setParam */
+  @Since("2.3.0")
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  @Since("2.3.0")
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  @Since("2.3.0")
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    transformSchema(dataset.schema, logging = true)
+    dataset.withColumn($(outputCol), func(col($(inputCol))))
+  }
+
+  @Since("2.3.0")
+  override def transformSchema(schema: StructType): StructType = {
+    func.inputTypes match {
+      case Some(funcInputType) =>
+        require(funcInputType.length == 1, "FuncTransformer only supports udf with one input")
+        val dataType = schema($(inputCol)).dataType
+        require(dataType == funcInputType.head, s"data type mismatch: udf input type" +
+          s" ${funcInputType.head}; inputCol ${$(inputCol)} data type $dataType ")
+      case None =>
+        val dataType = schema($(inputCol)).dataType
+        require(dataType.isInstanceOf[StructType], s"When func input types is None," +
+          s" FuncTransformer only supports StructType. ${$(inputCol)} is $dataType")
+    }
+    val outputFields = schema.fields :+ StructField($(outputCol), func.dataType, false)
+    StructType(outputFields)
+  }
+
+  @Since("2.3.0")
+  override def copy(extra: ParamMap): FuncTransformer = {
+    val copied = new FuncTransformer(uid, func)
+    copyValues(copied, extra)
+  }
+
+  @Since("2.3.0")
+  override def write: MLWriter = new FuncTransformerWriter(this)
+}
+
+/**
+ * :: Experimental ::
+ * Companion object for FuncTransformer with save and load function.
+ */
+@Experimental
+@Since("2.3.0")
+object FuncTransformer extends DefaultParamsReadable[FuncTransformer] {
+
+  private[FuncTransformer]
+  class FuncTransformerWriter(instance: FuncTransformer) extends MLWriter {
+
+    private case class Data(func: Array[Byte])
+
+    override protected def saveImpl(path: String): Unit = {
+      DefaultParamsWriter.saveMetadata(instance, path, sc)
+      val bo = new ByteArrayOutputStream()
+      new ObjectOutputStream(bo).writeObject(instance.func)
```

Comment:
Thanks @WeichenXu123 Those are good questions and at least we should cover the questions in the document.

1. I will test it, but perhaps with a small closure.
2. I imagine we'll still need the class dependency, yet IMO it's an expected behavior.

I'd like to learn if there's a better suggestion about how to save the udf object?